### PR TITLE
refactor(web): osk move & resize handler encapsulation 🍕

### LINKED ISSUE #5296: bug(web): resizing desktop OSK sometimes causes it to "pop" to a fixed width
**Describe the bug**

When you resize the desktop OSK, when you get larger than a certain size, it "pops" into a fixed, narrow width which is definitely incorrect.

image

See the screen capture for a live example: 

https://user-images.githubusercontent.com/4498365/122194512-db27ab80-ced8-11eb-991b-e76078d75ff2.mp4

**To Reproduce**

1. Go to https://keymanweb.com/#jv-java,Keyboard_jawa
2. Click in the text field and drag the corner of the OSK to resize it
3. See error

This may be related to the OSK sometimes loading at the same fixed width when loading a KeymanWeb page?

---

**KeymanWeb:**
 - OS: Windows 10 
 - Browser: Chrome 91 
 - KeymanWeb Version: 14.0.276

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #5296.

Continuing where #5393 left off, this PR addresses the "code-spaghetti" around the movement & resizing event handlers.  After a couple of development passes, it turns out that there's a fair bit of WETness between the move & resize handler sets... so I abstracted the WET aspects I could.

There are still a couple of sticky parts - the `uiManager.justActivated` bits, but those will be addressed later in the OSK-Core work.  We're not ready for the `onEventTriggeredFocus` aspect, which would allow the OSK-Core's consumer to handle that flag via event-raising, just yet.

### User Testing
@keymanapp/testers 

This PR is _all about_ desktop-form-factor OSK positioning and resizing.  That said, since this does attempt to extract said code from areas that _also_ affect the other form factors, some basic tests on other form factors would be wise, just to ensure no collateral damage happens.

#### Main focus

Please run the following tests on each of the following three browsers:
- Safari
- Chrome (on Windows)
- Firefox (on macOS, or Linux if feeling adventurous)

While the Chrome and Firefox tests could be run on the same OS, it's probably best to get a little cross-OS coverage in here as well.  If you want to be super-rigorous, test all possible OS-browser pairings... but that's probably overkill.

image

1. Dragging the title bar of the OSK should allow you to reposition it on the page.
    - When you do so, a "pin" symbol should appear:

image
    - Clicking that "pin" symbol should reset the OSK's position (and remove the "pin" symbol).

2. Dragging the "handle" at the bottom-right should allow you to resize the OSK.
    - Note:  there _are_ minimum and maximum limits.
    - Be sure to test an extremely tall and wide keyboard.  There should be no unwanted keyboard-size "snapping" effects like documented in #5296.

3.  Both resize _and_ reposition the OSK.  Then, refresh the page.  The OSK should "remember" where you left it and how large it should be after the refresh.

4.  With `sil_euro_latin`, while you should not be able to resize the OSK, you should be able to reposition it.

5.  Open the Developer console and enter the following command:
```
keyman.osk.setRect({'nomove': true, 'nosize': true, left: 0, top: 0})
```
  - You should no longer be able to move the OSK.
  - You should no longer be able to resize the OSK.
  - The OSK should snap to the top-left of the screen.

Now, enter this command, _without refreshing_:
```
keyman.osk.setRect({'nomove': false});
```
  - You should now be able to move, but not resize, the OSK.
  - Clicking the "pin" should reset the OSK's position, snapping it back to the top-left - its default position as set by the first call.
  - Before moving on, move the OSK away from the top-left position.

Now, enter this command, _without refreshing_:
```
keyman.osk.setRect({'nomove': true, 'nosize': false});
```
  - The OSK should remain where you left it, but will no longer be movable.
  - The OSK may now be resized.

Finally... refresh the page.  The OSK should resume default positioning behaviors, but will likely remember the previously set size.
    - If you had manually repositioned it, it should remember the last position it held.
    - Click the pin (which should have automatically appeared upon OSK load) and it should snap to the active element, not the top-left of the page.

#### Side-effect prevention

1.  Use Chrome mobile-device emulation for an iPhone X.
    - Verify that the OSK loads correctly.
    - Rotate the device; the OSK should rotate normally, same as before.
2. Load up the Android app from this branch and perform the same tests for a tablet form-factor.

So, one 'embedded' case and one not-embedded case for the mobile devices, and one with "phone" and one with "tablet".  If you want to be extra-rigorous, test all combinations, but I such rigor would be excessive here.

### CODE REVIEW

File: web/source/osk/layouts/mouseDragOperation.ts [hunk 1]
```diff
@@ -0,0 +1,188 @@
+namespace com.keyman.osk.layouts {
+ class CustomizationCoordinate {
```
Comment (mcdurdin) [inline web/source/osk/layouts/mouseDragOperation.ts:2]: Can you add a comment for this class?

File: web/source/osk/layouts/targetedFloatLayout.ts [hunk 2]
```diff
@@ -1,14 +1,214 @@
...
+            newWidth = 0.2*screen.width;
+          }
+          if(newHeight < 0.1*screen.height) {
+            newHeight = 0.1*screen.height;
+          }
+          if(newWidth > 0.9*screen.width) {
+            newWidth = 0.9*screen.width;
+          }
+          if(newHeight > 0.5*screen.height) {
+            newWidth = 0.5*screen.height;
```
Comment (mcdurdin) [inline web/source/osk/layouts/targetedFloatLayout.ts:179]:
```suggestion
            newHeight = 0.5*screen.height;
```

I am guessing this may fix #5296?

**Review by mcdurdin — COMMENTED** [review]
Can we get some user testing notes for this? Apart from that, I think the changes LGTM (and I have suggested a bug fix for an existing issue)

File: web/source/osk/layouts/targetedFloatLayout.ts (hunk 2 above)
Comment (jahorton) [reply web/source/osk/layouts/targetedFloatLayout.ts:179]: I knew it looked wrong, but didn't quite make that connection at the time, and it was the original code.  Good call.

Should be able to 🍒-pick this tidbit without much trouble.

Comment (jahorton) [reply web/source/osk/layouts/targetedFloatLayout.ts]: 🍒: #5428.

**Review by mcdurdin — APPROVED** [review]
LGTM pending user testing

**Comment by jahorton** [comment]
@keymanapp/testers

**Comment by MakaraSok** [comment]
**Safari 14.1.1 on macOS 11.4**
---

> 1. Dragging the title bar of the OSK should allow you to reposition it on the page.
> 2. Dragging the "handle" at the bottom-right should allow you to resize the OSK.
> 3. Both resize _and_ reposition the OSK.  Then, refresh the page.  The OSK should "remember" where you left it and how large it should be after the refresh.

**_result:_** all behave as expected.

> 4. With `sil_euro_latin`, while you should not be able to resize the OSK, you should be able to reposition it.

**_result:_** this behaves as expected.

> 5. Open the Developer console and enter the following command:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': true, left: 0, top: 0})
> ```
>   - You should no longer be able to move the OSK.
>   - You should no longer be able to resize the OSK.
>   - The OSK should snap to the top-left of the screen.

**_result:_** all behave as expected.

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': false});
> ```
>   - You should now be able to move, but not resize, the OSK.
>   - Clicking the "pin" should reset the OSK's position, snapping it back to the top-left - its default position as set by the first call.
>   - Before moving on, move the OSK away from the top-left position.

**_result:_** all behave as expected.

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': false});
> ```
>   - The OSK should remain where you left it, but will no longer be movable.
>   - The OSK may now be resized.

**_result:_** all behave as expected.
 
> Finally... refresh the page.  The OSK should resume default positioning behaviors, but will likely remember the previously set size.
>     - If you had manually repositioned it, it should remember the last position it held.
>     - Click the pin (which should have automatically appeared upon OSK load) and it should snap to the active element, not the top-left of the page.

**_result:_** all behave as expected.

image

------

**More test results coming up this afternoon ...**
---

**Comment by MakaraSok** [comment]
**Chrome 91.0.4472.114 (x86_64) on macOS 11.4**
---
> 1. Dragging the title bar of the OSK should allow you to reposition it on the page.
> 2. Dragging the "handle" at the bottom-right should allow you to resize the OSK.

**_result:_** all behave as expected here.

> 3. Both resize _and_ reposition the OSK.  Then, refresh the page.  The OSK should "remember" where you left it and how large it should be after the refresh.

**_result:_** the keyboard (English -English) was reset to "(System keyboard)" and resized/repositioned to default as shown below:

image


> 4. With `sil_euro_latin`, while you should not be able to resize the OSK, you should be able to reposition it.

**_result:_** the OSK can be resize and reposition, but there is an issue when trying -- reposition the OSK anywhere, type a key to see the OSK moves instantaneously to the default position. See the screen recording [here](https://drive.google.com/file/d/1KzSVJIV1FC4mEKdNSYv_F6abmNloc-oC/view?usp=sharing).

> 5. Open the Developer console and enter the following command:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': true, left: 0, top: 0})
> ```
>   - You should no longer be able to move the OSK.
>   - You should no longer be able to resize the OSK.
>   - The OSK should snap to the top-left of the screen.

**_result:_** everything behaves as expected.

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': false});
> ```
>   - You should now be able to move, but not resize, the OSK.
>   - Clicking the "pin" should reset the OSK's position, snapping it back to the top-left - its default position as set by the first call.
>   - Before moving on, move the OSK away from the top-left position.

**_result:_** everything behaves as expected.


> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': false});
> ```
>   - The OSK should remain where you left it, but will no longer be movable.
>   - The OSK may now be resized.

**_result:_** The OSK was moved to the bottom of the screen, but after running the command, the keyboard defaults to the top left corner while it is now not movable, but resizable. See the screeenshot below.
 
image

image

One noticeable behavior which may be out of the normal is that **the keyboard reset itself to the default size** after clicking away (i.e. in the console) and clicking back in the text box regardless of how big/small/tall/short the keyboard was set prior to clicking away.

> Finally... refresh the page.  The OSK should resume default positioning behaviors, but will likely remember the previously set size.
>     - If you had manually repositioned it, it should remember the last position it held.
>     - Click the pin (which should have automatically appeared upon OSK load) and it should snap to the active element, not the top-left of the page.

**_result:_** **Failed** - after the refresh, the keyboard moves to its default position and it didn't remember the size set before the refresh.

image

------

**Coming up next - Firefox 89.0.2 (x64) on macOS 11.4 ...**
---

**Comment by jahorton** [comment]
> ## **Chrome 91.0.4472.114 (x86_64) on macOS 11.4**

Interesting that you saw _any_ failures here - this is (essentially) the combination I used when developing the PR.  (Using Big Sur instead of Catalina, but that shouldn't make a difference.)  I went through and ran the specified tests again myself, to be sure, and could not reproduce the errors you reported here.

------

There's a chance may be one of those cases where Chrome needs a cache-clear in order to force it to pick up on a newly-compiled update from KMW.

image

A one-hour clear is usually enough to force Chrome to properly reload the KMW script.

So:
- make positively sure that KMW has been built since you last checked out the branch (rebuild KMW if not 100% sure)
- perform a Chrome cache-clear
- retest

(On occasion, similar measures may be needed within any browser.)

------

If the errors continue after that, I may need to schedule a video chat to see exactly what's happening.

**Comment by MakaraSok** [comment]
> > ## **Chrome 91.0.4472.114 (x86_64) on macOS 11.4**
> 
> Interesting that you saw _any_ failures here - this is (essentially) the combination I used when developing the PR. (Using Big Sur instead of Catalina, but that shouldn't make a difference.) I went through and ran the specified tests again myself, to be sure, and could not reproduce the errors you reported here.

Actions taken: (1) git pull from the branch again to be sure, (2) ran `./build.sh` and (3) clear cache in Chrome.

## **_Retest_ - Chrome 91.0.4472.114 (x86_64) on macOS 11.4**

> 1. Dragging the title bar of the OSK should allow you to reposition it on the page.

**_result:_** **OK**

> 2. Dragging the "handle" at the bottom-right should allow you to resize the OSK.

**_result:_** **OK**

> 3. Both resize _and_ reposition the OSK.  Then, refresh the page.  The OSK should "remember" where you left it and how large it should be after the refresh.

**_result:_** **failed** - see the behavior [here](https://drive.google.com/file/d/19n5zetpnVZTScKuoNOA8Y8seODRCuEso/view?usp=sharing)

> 4. With `sil_euro_latin`, while you should not be able to resize the OSK, you should be able to reposition it.

**_result:_** **failed** - the same behavior as shown [above](https://drive.google.com/file/d/1KzSVJIV1FC4mEKdNSYv_F6abmNloc-oC/view?usp=sharing)

> 5. Open the Developer console and enter the following command:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': true, left: 0, top: 0})
> ```
>   - You should no longer be able to move the OSK.
>   - You should no longer be able to resize the OSK.
>   - The OSK should snap to the top-left of the screen.

**_result:_** **OK**

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': false});
> ```
>   - You should now be able to move, but not resize, the OSK.
>   - Clicking the "pin" should reset the OSK's position, snapping it back to the top-left - its default position as set by the first call.
>   - Before moving on, move the OSK away from the top-left position.

**_result:_** **OK**

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': false});
> ```
>   - The OSK should remain where you left it, but will no longer be movable.

**_result:_** **failed** - the OSK snaps to the top-left after running the command - be informed that the OSK disappears completely from where it was moved to when clicking in the Console area to enter the command.

>   - The OSK may now be resized.

**_result:_** **OK**

> Finally... refresh the page.  The OSK should resume default positioning behaviors, but will likely remember the previously set size.
>     - If you had manually repositioned it, it should remember the last position it held.
>     - Click the pin (which should have automatically appeared upon OSK load) and it should snap to the active element, not the top-left of the page.

**_result:_** **failed** - after the refresh, the keyboard moves to its default position and it didn't remember the size set before the refresh

**Comment by MakaraSok** [comment]
**Chrome 91.0.4472.124 (x64) on Windows 10 Pro 21H1**
---

> 1. Dragging the title bar of the OSK should allow you to reposition it on the page.

**_result:_ Passed**

> 2. Dragging the "handle" at the bottom-right should allow you to resize the OSK.

**_result:_ Passed** - the handle does resize the OSK, but the issue [here](https://github.com/keymanapp/keyman/issues/5296) is experienced. 

> 3. Both resize _and_ reposition the OSK.  Then, refresh the page.  The OSK should "remember" where you left it and how large it should be after the refresh.

**_result:_ Passed**

> 4. With `sil_euro_latin`, while you should not be able to resize the OSK, you should be able to reposition it.

**_result:_ Passed** The OSK can be repositioned, but the "pin" is not in place to snap to the default position. (See the screenshot below)

image


> 5. Open the Developer console and enter the following command:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': true, left: 0, top: 0})
> ```
>   - You should no longer be able to move the OSK.
>   - You should no longer be able to resize the OSK.
>   - The OSK should snap to the top-left of the screen.

**_result:_ Passed**

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': false});
> ```
>   - You should now be able to move, but not resize, the OSK.
>   - Clicking the "pin" should reset the OSK's position, snapping it back to the top-left - its default position as set by the first call.
>   - Before moving on, move the OSK away from the top-left position.

**_result:_ Passed**

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': false});
> ```
>   - The OSK should remain where you left it, but will no longer be movable.
>   - The OSK may now be resized.

**_result:_ Passed**


> Finally... refresh the page.  The OSK should resume default positioning behaviors, but will likely remember the previously set size.
>     - If you had manually repositioned it, it should remember the last position it held.
>     - Click the pin (which should have automatically appeared upon OSK load) and it should snap to the active element, not the top-left of the page.

**_result:_ Passed**

---

**Firefox 89.0.2 (x64) on Windows 10 Pro 21H1**
---

> 1. Dragging the title bar of the OSK should allow you to reposition it on the page.

**_result:_ Passed**

> 2. Dragging the "handle" at the bottom-right should allow you to resize the OSK.

**_result:_ Passed** the handle does the job, but the issue (bug(web): resizing desktop OSK sometimes causes it to "pop" to a fixed width #5296) is still observed, just like that of Chrome.

> 3. Both resize _and_ reposition the OSK.  Then, refresh the page.  The OSK should "remember" where you left it and how large it should be after the refresh.

**_result:_ Passed**

> 4. With `sil_euro_latin`, while you should not be able to resize the OSK, you should be able to reposition it.

**_result:_ Passed** - but the pin on the top right is not in place to snap back to the default position. This issue has also been experienced in Chrome (see above).

> 5. Open the Developer console and enter the following command:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': true, left: 0, top: 0})
> ```
>   - You should no longer be able to move the OSK.
>   - You should no longer be able to resize the OSK.
>   - The OSK should snap to the top-left of the screen.

**_result:_ Failed** - the OSK doesn't snap to the top-left of the screenp it stays where it was before the command was executed.

image


> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': false});
> ```
>   - You should now be able to move, but not resize, the OSK.
>   - Clicking the "pin" should reset the OSK's position, snapping it back to the top-left - its default position as set by the first call.
>   - Before moving on, move the OSK away from the top-left position.

**_result:_ Passed**

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': false});
> ```
>   - The OSK should remain where you left it, but will no longer be movable.
>   - The OSK may now be resized.

**_result:_ Passed**


> Finally... refresh the page.  The OSK should resume default positioning behaviors, but will likely remember the previously set size.
>     - If you had manually repositioned it, it should remember the last position it held.
>     - Click the pin (which should have automatically appeared upon OSK load) and it should snap to the active element, not the top-left of the page.

**_result:_ Passed**

-------

> #### Side-effect prevention
> 
> 1.  Use Chrome mobile-device emulation for an iPhone X.
>     - Verify that the OSK loads correctly.
>     - Rotate the device; the OSK should rotate normally, same as before.
> 2. Load up the Android app from this branch and perform the same tests for a tablet form-factor.
> 
> So, one 'embedded' case and one not-embedded case for the mobile devices, and one with "phone" and one with "tablet".  If you want to be extra-rigorous, test all combinations, but I such rigor would be excessive here.

pending...

**Comment by jahorton** [comment]
Huh.  Now that I found the reason Firefox wasn't acting how I expected... how did it take until Firefox for that issue to pop up?

Pushed a fix for it, at least.

**Comment by MakaraSok** [comment]
> Huh. Now that I found the reason Firefox wasn't acting how I expected... how did it take until Firefox for that issue to pop up?
> 
> Pushed a fix for it, at least.

Would you like another round of testing for Chrome and FireFox on Windows? The macOS ones are out of reach now given that the browsers are not behaving as expected. :(

**Comment by jahorton** [comment]
Yes please.

**Comment by MakaraSok** [comment]
**Retest: Chrome 91.0.4472.164 (x64) on Windows 10 Pro 21H1**
---
On the master branch, after `git pull` the following commands were executed:
```
git fetch origin
git checkout -b refactor/web/osk-header-footer-encapsulation origin/refactor/web/osk-header-footer-encapsulation
```
Then, `./build.sh` from within `~/keyman/web/source`.

The version of Chrome browser was used to access: `http://localhost/keymanweb/testing/unminified.html`.

**Here are the results of each**

> 1. Dragging the title bar of the OSK should allow you to reposition it on the page.

**_result:_ Passed**

> 2. Dragging the "handle" at the bottom-right should allow you to resize the OSK.
>      - Note: there are minimum and maximum limits.
>      - Be sure to test an extremely tall and wide keyboard. There should be no unwanted keyboard-size "snapping" effects like documented in bug(web): resizing desktop OSK sometimes causes it to "pop" to a fixed width #5296.

**_result:_ The snapping effects are experienced when dragging the handle to the far right of the screen.** 

> 3. Both resize _and_ reposition the OSK.  Then, refresh the page.  The OSK should "remember" where you left it and how large it should be after the refresh.

**_result:_ Passed**

> 4. With `sil_euro_latin`, while you should not be able to resize the OSK, you should be able to reposition it.

**_result:_ The OSK can be repositioned, but the "pin" is not in place to snap to the default position.**

> 5. Open the Developer console and enter the following command:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': true, left: 0, top: 0})
> ```
>   - You should no longer be able to move the OSK.
>   - You should no longer be able to resize the OSK.
>   - The OSK should snap to the top-left of the screen.

**_result:_ Failed -- The OSK doesn't snap to the top-left of the screen. See the screenshot below:**

image


> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': false});
> ```
>   - You should now be able to move, but not resize, the OSK.
>   - Clicking the "pin" should reset the OSK's position, snapping it back to the top-left - its default position as set by the first call.
>   - Before moving on, move the OSK away from the top-left position.

**_result:_ Passed**

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': false});
> ```
>   - The OSK should remain where you left it, but will no longer be movable.
>   - The OSK may now be resized.

**_result:_ Passed**


> Finally... refresh the page.  The OSK should resume default positioning behaviors, but will likely remember the previously set size.
>     - If you had manually repositioned it, it should remember the last position it held.
>     - Click the pin (which should have automatically appeared upon OSK load) and it should snap to the active element, not the top-left of the page.

**_result:_ Passed**

---

**Firefox 90.0 (x64) on Windows 10 Pro 21H1**
---

The version of Firefox browser was used to access: `http://localhost/keymanweb/testing/unminified.html`.

**Here are the results of each**

> 1. Dragging the title bar of the OSK should allow you to reposition it on the page.

**_result:_ Passed**

> 2. Dragging the "handle" at the bottom-right should allow you to resize the OSK.

**_result:_ Passed** 

> 3. Both resize _and_ reposition the OSK.  Then, refresh the page.  The OSK should "remember" where you left it and how large it should be after the refresh.

**_result:_ Passed**

> 4. With `sil_euro_latin`, while you should not be able to resize the OSK, you should be able to reposition it.

**_result:_ Passed**

> 5. Open the Developer console and enter the following command:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': true, left: 0, top: 0})
> ```
>   - You should no longer be able to move the OSK.
>   - You should no longer be able to resize the OSK.
>   - The OSK should snap to the top-left of the screen.

**_result:_ Passed**

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': false});
> ```
>   - You should now be able to move, but not resize, the OSK.
>   - Clicking the "pin" should reset the OSK's position, snapping it back to the top-left - its default position as set by the first call.
>   - Before moving on, move the OSK away from the top-left position.

**_result:_ Passed**

> Now, enter this command, _without refreshing_:
> ```
> keyman.osk.setRect({'nomove': true, 'nosize': false});
> ```
>   - The OSK should remain where you left it, but will no longer be movable.
>   - The OSK may now be resized.

**_result:_ Passed**

> Finally... refresh the page.  The OSK should resume default positioning behaviors, but will likely remember the previously set size.
>     - If you had manually repositioned it, it should remember the last position it held.
>     - Click the pin (which should have automatically appeared upon OSK load) and it should snap to the active element, not the top-left of the page.

**_result:_ Passed**

-------

> #### Side-effect prevention
> 
> 1.  Use Chrome mobile-device emulation for an iPhone X.
>     - Verify that the OSK loads correctly.
>     - Rotate the device; the OSK should rotate normally, same as before.

**_result:_** - the keyboard doesn't shown up until the tapping in the text area. 

> 2. Load up the Android app from this branch and perform the same tests for a tablet form-factor.

**_result:_** - there is a black bar on top of the keyboard when switch to horizontal:

- Emulator Pixel_C_API_22, Android 5.1.1

Screenshot_1626075654

- Emulator Pixel_C_API_29, Android 10

vertical:
image

horizontal:
image

horizontal - a keyboard without predictive text also shows the black bar
Screenshot_1626667056




- Emulator Pixel_2_API_29, Android 10 -

Screenshot_1626078855


> 
> So, one 'embedded' case and one not-embedded case for the mobile devices, and one with "phone" and one with "tablet".  If you want to be extra-rigorous, test all combinations, but I such rigor would be excessive here.

**_need a GVC on this_**

**Comment by jahorton** [comment]
> need a GVC on this

A "GVC"?

**Comment by MakaraSok** [comment]
A call, probably after today's standup.

**Comment by jahorton** [comment]
... how did I not notice the black bar issue from the previous error report until now?  Just realized this isn't the first testing round from which you mentioned that.

At any rate, I don't see it in the iOS app or its system keyboard.  I can reproduce this via Android emulation, so I'll pursue it that way.

**Comment by jahorton** [comment]
Uh... just to be sure, I checked that behavior against `master`.  Unless I need to refresh something... that black bar bit's a bug on our current `master`... and has been for a while.

I've managed to trace the bug back to [`15.0.63`](https://github.com/keymanapp/keyman/blob/master/HISTORY.md#15063-alpha-2021-06-08), at least as it reproduces on my machine + emulator setup.  That screams #5201 to me, given the history.md entries for that release.

**Comment by jahorton** [comment]
All of the other test results in the most recent report above are as expected.  (Given known limitations of the various browsers as mentioned in other comments.)

**Comment by mcdurdin** [comment]
> That screams #5201 to me, given the history.md entries for that release.

See also #5252 (coming out of that) which sounds related.

**Comment by jahorton** [comment]
Yeah, that does describe it pretty well.

... I totally should have checked the "known issues" list instead of running multiple builds to diagnose.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.85-alpha](https://keyman.com/downloads/releases/alpha/15.0.85)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.88-alpha](https://keyman.com/downloads/releases/alpha/15.0.88)

**Pull request merged**